### PR TITLE
8336640: Shenandoah: Parallel worker use in parallel_heap_region_iterate

### PR DESCRIPTION
**Pull request by pengxiaolong**
[parallel_heap_region_iterate](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp#L1726-L1734) is used to execute lightweight operations on heap regions, including ShenandoahPrepareForMarkClosure, ShenandoahInitMarkUpdateRegionStateClosure, ShenandoahFinalUpdateRefsUpdateRegionStateClosure, ShenandoahResetUpdateRegionStateClosure and ShenandoahFinalMarkUpdateRegionStateClosure. Since all the operations are very lightweight, in regular cases w/o large number of heap regions, the parallelism seems to be an overkill because the cost of multi-thread orchestrating could be more expensive; In most cases, single thread should be more efficient. Also, if multiple threading is needed, we should maximize the utilization of all active workers for best performance.

This PR includes proposed improvments addressing the known issues:
1. Change the default value of ShenandoahParallelRegionStride to 0, when it is 0, Shenandoah will auto derive the value of stride for best performance; 
2. if num_regions is <= 4096, not use worker threads at all to avoid the overhead of multi-threading;
3. When num_regions is more than 4096, use worker threads to parallelize the workload, derive the value of stride to evenly distribute the workload to all active workers.
4. When number of active workers is 1, don't bother the workers, it is faster to finish the workload in current thread(avoid overhead of multi-threads orchestration)

There are some time metrics I collected from test with TIP version(I added time metrics for parallel_heap_region_iterate):

JVM args: export JAVA_OPTS="-Xms8G -Xmx8G  -XX:+AlwaysPreTouch -XX:+UseShenandoahGC -XX:+UnlockExperimentalVMOptions -XX:ShenandoahParallelRegionStride=<stride> -XX:ShenandoahTargetNumRegions=<num_regions>  -Xlog:gc*"

|             | 1024 regions | 2048 regions | 4096 regions | 8192 regions |16384 regions |
| ----------- | ------------ | ------------ | ------------ | ------------ |------------ |
| 1024 stride | 5785 ns         | 22194 ns        | 20953 ns        | 23008 ns        |33013 ns       |
| 2048 stride | N/A          | 6491 ns         | 22476  ns        | 25842 ns        |34378 ns        |
| 4096 stride | N/A          | N/A          | 14034 ns        | 28425 ns        |36324 ns        |
| 8192 stride | N/A          | N/A          | N/A          | 24359 ns        |45231 ns        |
| 16384 stride | N/A          | N/A          | N/A          | N/A          |53679 ns        |

Basically when we increase stride, less threads are used for parallel iteration, we get worse latency which is expected. when number of regions is same as stride, it won't use mutli-threading, using single thread to process 4096 regions is much better then 4 threads(1024 stride). 

For the PR, also tested with following JVM args:
export JAVA_OPTS="-Xms8G -Xmx8G  -XX:+AlwaysPreTouch -XX:+UseShenandoahGC -XX:+UnlockExperimentalVMOptions  -XX:ShenandoahTargetNumRegions=<num_regions>  -Xlog:gc*"

| Regions | time (ns) |
| ------- | --------- |
| 1024    | 5103      |
| 2048    | 6132      |
| 4096    | 12763     |
| 8192    | 24295     |
| 16384    | 33729     |

Overall the performance is optimal no matter how many heap regions.

Additional test:
- [x] `make test TEST=hotspot_gc_shenandoah`
```
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/hotspot/jtreg:hotspot_gc_shenandoah      259   259     0     0   
==============================
TEST SUCCESS
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336640](https://bugs.openjdk.org/browse/JDK-8336640): Shenandoah: Parallel worker use in parallel_heap_region_iterate (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20305/head:pull/20305` \
`$ git checkout pull/20305`

Update a local copy of the PR: \
`$ git checkout pull/20305` \
`$ git pull https://git.openjdk.org/jdk.git pull/20305/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20305`

View PR using the GUI difftool: \
`$ git pr show -t 20305`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20305.diff">https://git.openjdk.org/jdk/pull/20305.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20305#issuecomment-2248724614)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 1]
```diff
@@ -1725,8 +1726,18 @@ class ShenandoahParallelHeapRegionTask : public WorkerTask {
...
-    ShenandoahParallelHeapRegionTask task(blk);
+  const uint active_workers = workers() -> active_workers();
+  const size_t n_regions = num_regions();
+  size_t stride = ShenandoahParallelRegionStride;
+  if (stride == 0 && active_workers > 1) {
+    // Automatically derive the stride for ShenandoahParallelHeapRegionTask if ShenandoahParallelRegionStride is 0.
+    // There is overhead to orchastrate the worker threads, when num_regions is less than 4096 set stride to 4096 to
+    // not use worker threads to avoid the overhead; otherwise cacluate the stride by num_regions/active_workers
+    // to make sure every worker thread will have same amount of workload.
+    stride = n_regions <= 4096 ? 4096 : checked_cast<size_t>(ceil(checked_cast<float>(n_regions) / checked_cast<float>(active_workers)));
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1737]: I suggest writing it like this:

```
  size_t stride = ShenandoahParallelRegionStride;

  if (stride == 0 && active_workers > 1) {
    // Automatically derive the stride to balance the work between threads
    // evenly. Do not try to split work if below the reasonable threshold.
    const size_t threshold = 4096;
    stride = (n_regions <= threshold) ?
            threshold :
            (n_regions + active_workers - 1) / active_workers;
  }

  if (n_regions > stride && active_workers > 1) {
```

Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1737]: Neat! Thanks! I'll update the PR.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 2]
```diff
@@ -1694,19 +1694,20 @@ class ShenandoahParallelHeapRegionTask : public WorkerTask {
 private:
   ShenandoahHeap* const _heap;
   ShenandoahHeapRegionClosure* const _blk;
+  size_t _stride;
```
[the remark this answers is not in the document]
Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1697]: Yes, it should be const

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp [hunk 3]
```diff
@@ -318,9 +318,10 @@
           "checking for cancellation, yielding, etc. Larger values improve "\
           "marking performance at expense of responsiveness.")              \
                                                                             \
-  product(uintx, ShenandoahParallelRegionStride, 1024, EXPERIMENTAL,        \
+  product(uintx, ShenandoahParallelRegionStride, 0, EXPERIMENTAL,           \
           "How many regions to process at once during parallel region "     \
-          "iteration. Affects heaps with lots of regions.")                 \
+          "iteration. Affects heaps with lots of regions."                  \
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp:323]: Probably want a trailing space after `regions. "`

Comment (pengxiaolong) [reply src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp]: Thanks! Just added.

**Review by shipilev — APPROVED** [review]

**Review by earthling-amzn — APPROVED** [review]
Nice work!

**Review by ysramakrishna — APPROVED** [review]

**Comment by pengxiaolong** [comment]
Thank you all for the reviews!  I'll integrate the PR.

**Pull request closed without merging**